### PR TITLE
Add more tests to tiltPlot functions

### LINKED ISSUE #114: Create more tests for my functions
Motivation: My new function of the scatter plot breaks my other function's code. Write new tests that makes my code less easy to break.

Examples : 
- Create edge cases for missing values: which columns is crucial not to have NA's in them ? If we have NA's in any of them, how should we handle them ?
- Proportions : goes up to 1
- Several benchmarks are plotted if the users choose several benchmarks (horizontal bar plots, ...)
- Amount_total is the correct one in my plots
- Every mode chosen work for the users
- ...

### PR DESCRIPTION
**Pull request by lindadelacombaz**
Closes #114.
Dear @maurolepore,
Here is my PR that add more precise tests to my functions.

Thank you ! :)
----

TODO

- [x] [Link related issue/PR]([url](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)). 
- [x] Describe the goal of the PR. Avoid details that are clear in the diff.
- [x] Mark the PR as draft.
- [x] [Include a unit test](https://code-review.tidyverse.org/reviewer/aspects.html#sec-tests).
- [x] Review your own PR in "Files changed".
- [x] Ensure the PR branch is updated.
- [x] Ensure the checks pass.
- [x] Change the status from draft to ready.
- [x] Polish the PR title and description.
- [x] Assign a reviewer.

EXCEPTIONS

- [ ] Slide here any item that you intentionally choose to not do.

### CODE REVIEW

File: R/utils-test.R [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#' example_financial()
+example_financial <- function(bank_id = "a",
+                              amount_total = 10,
+                              company_name = "b",
+                              emission_profile = "medium",
+                              benchmark = "c",
+                              equal_weight_finance = 10,
+                              worst_case_finance = 10,
+                              best_case_finance = 10,
+                              ...) {
```
Comment (maurolepore) [inline R/utils-test.R:19]: Great! This is exactly how I would do it. 

To avoid littering your tests with brittle column names, you may want to support creating example datasets with your `example_*()` function in combination with stable interfaces that hide those brittle column names. I show this in the reprex. Let me know if it's unclear.

<details/>
<summary/>
reprex
</summary>

``` r
  library(rlang, warn.conflicts = FALSE)
  library(tibble, warn.conflicts = FALSE)

  # R/example_datasets.R -------------------------------------------------------
  
  # Idea used in tiltIndicator::example_companies() and friends
  # You need to add `:=` with this: `usethis::use_import_from("rlang", ":=")`
  example_financial <- function (...)  {
    new <- tibble(...)
    old <- as.list(example_financial_impl())
    add <- setdiff(names(old), names(new))
    as_tibble(append(new, old[add]))
  }
  # This is exactly what you wrote but without the dots `...` 
  example_financial_impl <- function(bank_id = 1, other = 1) {
    tibble(
      bank_id = bank_id,
      other = other
    )
  }
  
  # This file mentions brittle names, so if it changes you will also need to change
  # column names everywhere else, unless everywhere you access those brittle
  # names through stable interfaces that YOU control. So it seems like a good
  # place to map those brittle names to your stable interfaces
  get_id <- function() "bank_id"
  get_europages_product <- function() "ep_product"
  
  
  
  # Usage -----------------------------------------------------------------------
  
  # This version of `example_financial()` supports the basic behaviour ...
  example_financial(bank_id = 1:2)
#> # A tibble: 2 × 2
#>   bank_id other
#>     <int> <dbl>
#> 1       1     1
#> 2       2     1
  example_financial(bank_id = NULL)
#> # A tibble: 1 × 2
#>   bank_id other
#>     <dbl> <dbl>
#> 1       1     1
  example_financial(new = 1)
#> # A tibble: 1 × 3
#>     new bank_id other
#>   <dbl>   <dbl> <dbl>
#> 1     1       1     1
  
  # But it also supports using your stable `get_*()` functions
  example_financial(!!get_id() := letters[1:3])
#> # A tibble: 3 × 2
#>   bank_id other
#>   <chr>   <dbl>
#> 1 a           1
#> 2 b           1
#> 3 c           1
  
  example_financial(!!get_id() := letters[1:3], !!get_europages_product() := 1)
#> # A tibble: 3 × 3
#>   bank_id ep_product other
#>   <chr>        <dbl> <dbl>
#> 1 a                1     1
#> 2 b                1     1
#> 3 c                1     1
```

</details>

File: tests/testthat/test-bar_plot_emission_profile_financial.R [hunk 2]
```diff
@@ -1,20 +1,36 @@
 test_that("returns an object of the expected class", {
-  plot <- bar_plot_emission_profile_financial(financial, benchmarks())
+  data <- example_financial(!!aka("europages_product") := "e")
+  plot <- bar_plot_emission_profile_financial(data, benchmarks())
```
Comment (maurolepore) [inline tests/testthat/test-bar_plot_emission_profile_financial.R:3]: @lindadelacombaz here's how you can now use the powers of `example_financial()`

File: R/utils.R [hunk 3]
```diff
@@ -131,7 +131,8 @@ dictionary <- function() {
         "emission_profile",      "emission_profile",
          "profile_ranking",       "profile_ranking",
    "transition_risk_score", "transition_risk_score",
-         "transition_risk",       "transition_risk"
+         "transition_risk",       "transition_risk",
+       "europages_product",            "ep_product",
```
Comment (maurolepore) [inline R/utils.R:135]: Added this stuff.

File: NAMESPACE [hunk 4]
```diff
@@ -4,7 +4,9 @@ export(bar_plot_emission_profile)
 export(bar_plot_emission_profile_financial)
 export(benchmarks)
 export(example_financial)
+export(example_without_financial)
```
Comment (maurolepore) [inline NAMESPACE:7]: SELF Do we need to export it?

File: tests/testthat/test-map_region_risk.R [hunk 5]
```diff
@@ -1,27 +1,86 @@
...
-    plot <- map_region_risk(without_financial)
-  })
+  skip_on_ci()
+  data <- example_without_financial(
+    postcode = 53773L,
+    !!aka("emission_profile") := c("high", "medium", "low")
+  )
+  plot <- map_region_risk(data)
   expect_s3_class(plot, "ggplot")
 })
```
Comment (maurolepore) [inline tests/testthat/test-map_region_risk.R:9]: `map_region_risk()` is terribly slow. You may find a way to mock the call to the external API (assuming that's the bottleneck). But in the meantime you may want to minimize the number of tests you run on this function -- by removing redundant tests like this one. Don't worry, if your function doesn't return a ggplot object, your other tests will let you know.

<details/>
<summary/>
reprex
</summary>

``` r
devtools::load_all()
#> ℹ Loading tiltPlot

# Relatively fast
system.time(plot_sankey(financial))
#>    user  system elapsed 
#>   0.051   0.000   0.051
# Relatively slow (actually very slow)
system.time(map_region_risk(without_financial))
#> Extracting data using giscoR package, please report issues on https://github.com/rOpenGov/giscoR/issues
#> Cache management as per giscoR. see 'giscoR::gisco_get_nuts()'
#>    user  system elapsed 
#>   1.137   0.081  14.600
```

</details>

File: tests/testthat/test-map_region_risk.R [hunk 6]
```diff
@@ -1,27 +1,86 @@
...
+    postcode = c(53773L, 53774L, 53775L),
+    !!aka("emission_profile") := c("high", "medium", "low")
+  )
+  plot <- map_region_risk(data)
+
+  postcodes <- unique(plot$plot_env$aggregated_data$postcode)
+  expected_postcodes <- unique(data$postcode)
+
+  expect_equal(sort(postcodes), sort(expected_postcodes))
+})
```
Comment (maurolepore) [inline tests/testthat/test-map_region_risk.R:42]: See the humble-object pattern above.

File: tests/testthat/test-scatter_plot_financial.R [hunk 7]
```diff
@@ -1,36 +1,97 @@
...
+    !!aka("tilt_sector") := "t"
+  )
+
+  lapply(modes(), function(mode) {
+    test_rank(data, mode, aka("profile_ranking"))
+  })
+  lapply(modes(), function(mode) {
+    test_rank(data, mode, aka("emission_profile"))
+  })
+})
```
Comment (maurolepore) [inline tests/testthat/test-scatter_plot_financial.R:72]: This allows you to run many tests with little code. But see my comments above. In tests sometimes it's best to focus on fewer, more meaningful tests than lots' of overlapping ones.

**Review by maurolepore — COMMENTED** [review]
Thanks Linda. I wrote some comments to think about.
